core/src/main/kotlin/com/strumenta/kolasu/model/Model.kt

```diff
@@ -1,18 +1,27 @@
 package com.strumenta.kolasu.model
 
-import com.strumenta.kolasu.mapping.position
-import org.antlr.v4.runtime.ParserRuleContext
+interface WithPosition {
+    val position: Position?
+}
+
+class JustPosition(override val position: Position) : WithPosition
 
 /**
  * The Abstract Syntax Tree will be constituted by instances of Node.
  */
-open class Node(
-    /**
-     * Explicitly set position.
-     * @see Node.position
-     */
-    open var specifiedPosition: Position? = null
-) {
+open class Node() : WithPosition {
+
+    constructor(position: Position?) : this() {
+        if (position != null) {
+            origin = JustPosition(position)
+        }
+    }
+
+    constructor(origin: WithPosition?) : this() {
```

Comment:
Do all origins have a position? What if I create nodes arbitrarily? Would it be useful to have a type named Origin?

Reply:
Would it make sense to have origin has something more generic, so that it can indicate not only origin through parsing but alternatively something like "GeneratedByProgramX" or "Obtained By transforming  AST node Y" ?

Reply:
In Origin, the position is optional, so yes indeed renaming WithPosition to Origin would be appropriate